src/chain/optimism.rs

```diff
@@ -0,0 +1,286 @@
+        return Err(OptimismTransactionParsingError::InvalidType(tx_type_raw));
+    };
+
+    match tx_type {
+        OpTxType::Legacy | OpTxType::Eip2930 => tx
+            .inner
+            .gas_price
+            .map(U256::from)
+            .ok_or(OptimismTransactionParsingError::MissingGasPrice),
+        OpTxType::Eip1559 | OpTxType::Eip4844 => tx
+            .inner
+            .max_fee_per_gas
+            .map(U256::from)
+            .ok_or(OptimismTransactionParsingError::MissingMaxFeePerGas),
+        OpTxType::Deposit => Ok(U256::ZERO),
+    }
+}
+
+/// Convert [Transaction] to [OptimismFields]
+pub(crate) fn get_optimism_fields(
+    tx: &Transaction,
+) -> Result<OptimismFields, OptimismTransactionParsingError> {
+    let envelope_buf = {
+        let tx_type = tx.inner.transaction_type.unwrap_or_default();
+        let op_tx_type = OpTxType::try_from(tx_type)
+            .map_err(|_err| OptimismTransactionParsingError::UnexpectedType(tx_type))?;
+        let inner = tx.inner.clone();
+        let tx_envelope = match op_tx_type {
+            OpTxType::Legacy => Signed::<TxLegacy>::try_from(inner).map(OpTxEnvelope::from),
+            OpTxType::Eip2930 => Signed::<TxEip2930>::try_from(inner).map(OpTxEnvelope::from),
+            OpTxType::Eip1559 => Signed::<TxEip1559>::try_from(inner).map(OpTxEnvelope::from),
+            OpTxType::Eip4844 => Signed::<TxEip4844>::try_from(inner).map(OpTxEnvelope::from),
+            OpTxType::Deposit => {
+                let tx_deposit = TxDeposit {
+                    source_hash: tx
+                        .source_hash
+                        .ok_or(OptimismTransactionParsingError::MissingSourceHash)?,
+                    from: tx.inner.from,
+                    to: tx.inner.to.into(),
+                    mint: tx.mint,
+                    value: tx.inner.value,
+                    gas_limit: tx.inner.gas,
+                    is_system_transaction: tx.is_system_tx.unwrap_or_default(),
+                    input: tx.inner.input.clone(),
+                };
+                Ok(OpTxEnvelope::from(tx_deposit))
+            }
+        }
+        .map_err(|err| OptimismTransactionParsingError::ConversionError(err.to_string()))?;
+
+        let mut envelope_buf = Vec::<u8>::new();
+        tx_envelope.encode_2718(&mut envelope_buf);
+        Bytes::from(envelope_buf)
+    };
+
+    Ok(OptimismFields {
+        source_hash: tx.source_hash,
+        mint: tx.mint,
+        is_system_transaction: tx.is_system_tx,
+        enveloped_tx: Some(envelope_buf),
+    })
+}
+
+impl PevmChain for PevmOptimism {
+    type Transaction = Transaction;
+    type BlockSpecError = OptimismBlockSpecError;
+    type TransactionParsingError = OptimismTransactionParsingError;
+
+    fn id(&self) -> u64 {
+        self.id
+    }
+
+    fn build_tx_from_alloy_tx(&self, tx: alloy_rpc_types::Transaction) -> Self::Transaction {
+        Transaction {
+            inner: tx,
+            ..Transaction::default()
```

Comment:
While this is okay for testing, I wonder if this works for real usage. Maybe it's safer to list the remaining fields with a comment why these can be default.

Reply:
The remaining fields are `Option<_>`, which are always `None` unless the transaction is a Deposit Transaction (tx type = 126).

So, the fn implementation is realistic.

Reply:
How is it realistic when you just name a case it handles wrong.. We should at least add a `TODO` to address that, or note why we don't support it.

Reply:
`alloy_rpc_types::Transaction` can never express a Deposit Transaction, because it does not have the dedicated OP fields (`mint`, `source_hash`, `is_system_tx`, `deposit_receipt_version`).

In other words, there is no way a user can (accidentally or intentionally) create a Deposit Transaction of type `alloy_rpc_types::Transaction`.

Therefore, a `TODO:` is not needed.

-----

We can expand the `..Transaction::default()`, just to make it more verbose.

Reply:
The code will look like this:

```
fn build_tx_from_alloy_tx(&self, tx: alloy_rpc_types::Transaction) -> Self::Transaction {
        Self::Transaction {
            inner: tx,
            mint: None,
            source_hash: None,
            is_system_tx: None,
            deposit_receipt_version: None,
        }
    }

```

Reply:
Okay, let's keep the current code and add a `TODO` to support building a deposit transaction. For instance, by: 
- Add a `TxExtraFields` type to `PevmChain` with a `Default` constraint. It is `()` for Ethereum and a struct of four fields for OP.
- Convert this function to `fn build_tx(&self, base: alloy_rpc_types::Transaction, extra: Self::TxExtraFields) Self::Transaction`.

It is of low priority until we write tests with deposits, even then we can construct them manually.